x-pack/filebeat/input/http_endpoint/gzip.go

```diff
@@ -18,29 +19,34 @@ var gzipDecoderPool = sync.Pool{
 
 type pooledGzipReader struct {
 	Reader *gzip.Reader
+	closer io.Closer
 }
 
-func newPooledGzipReader(r io.Reader) (*pooledGzipReader, error) {
+func newPooledGzipReader(r io.ReadCloser) (*pooledGzipReader, error) {
 	gzipReader := gzipDecoderPool.Get().(*gzip.Reader)
 	if err := gzipReader.Reset(r); err != nil {
 		gzipDecoderPool.Put(gzipReader)
 		return nil, err
 	}
-	return &pooledGzipReader{Reader: gzipReader}, nil
+	return &pooledGzipReader{Reader: gzipReader, closer: r}, nil
 }
 
 // Read implements io.Reader, reading uncompressed bytes from its underlying Reader.
 func (r *pooledGzipReader) Read(b []byte) (int, error) {
 	return r.Reader.Read(b)
 }
 
-// Close closes the Reader. It does not close the underlying io.Reader.
+// Close closes the Reader and the underlying source.
 // In order for the GZIP checksum to be verified, the reader must be
 // fully consumed until the io.EOF.
 //
 // After this call the reader should not be reused because it is returned to the pool.
 func (r *pooledGzipReader) Close() error {
 	err := r.Reader.Close()
+	_err := r.closer.Close()
+	if err == nil {
+		err = errors.Join(err, _err)
```

Comment:
Should we ignore `error` from `closer.Close()` when `Reader.Close()` returns an error?

Reply:
We don't, though the logic here can be simplified (what I have is a Frankenstein's monster that I should clean up). Thank you for raising it.